File: algorithms/python/677/map_sum_pairs.py
from typing import List


class Item:
    def __init__(self, key: str, value: int):
        self.key: str = key
        self.value: int = value

    def __repr__(self) -> str:
        return f"<Item: key: {self.key}, value: {self.value}>"


def bisect_left(a: List[Item], x: str) -> int:
    lo, hi = 0, len(a)
    while lo < hi:
        mid = (lo + hi) // 2
        if a[mid].key < x:
            lo = mid + 1
        else:
            hi = mid
    return lo


class MapSum:

    def __init__(self):
        self.items: List[Item] = []

    def search(self, key: str) -> int:
        return bisect_left(self.items, key)

    def insert(self, key: str, val: int) -> None:
        idx = self.search(key)
        if idx == len(self.items):
            self.items.append(Item(key, val))
            return

        if self.items[idx].key == key:
            self.items[idx].value = val
            return

        self.items.append(Item(key, val))
        self.items.sort(key=lambda x: x.key)

    def sum(self, prefix: str) -> int:
        res = 0
        for i in range(self.search(prefix), len(self.items)):
            if not self.items[i].key.startswith(prefix):
                break
            res += self.items[i].value
        return res


if __name__ == "__main__":
    m = MapSum()
    assert m.sum("ap") == 0
    m.insert("apple", 10)
    assert m.sum("bap") == 0
    assert m.sum("ap") == 10
    m.insert("apple", 20)
    assert m.sum("ap") == 20, f"got: {m.sum('ap')}"
    m.insert("bb", 10)
    m.insert("app", 1)
    assert m.sum("ap") == 21

File: algorithms/go/677/map_sum_pairs.go
package main

import (
	"sort"
	"strings"
)

type Item struct {
	key   string
	value int
}

type Items []Item

func (items Items) Len() int {
	return len(items)
}

func (items Items) Less(i, j int) bool {
	if items[i].key < items[j].key {
		return true
	}
	return false
}

func (items Items) Swap(i, j int) {
	items[i], items[j] = items[j], items[i]
}

type MapSum struct {
	items Items
}

func Constructor() MapSum {
	return MapSum{Items{}}
}

func (this *MapSum) Insert(key string, val int) {
	idx := sort.Search(len(this.items), func(i int) bool {
		return this.items[i].key >= key
	})
	if idx == len(this.items) {
		this.items = append(this.items, Item{key, val})
		return
	}
	if this.items[idx].key == key {
		this.items[idx].value = val
		return
	}

	this.items = append(this.items, Item{key, val})
	sort.Sort(this.items)

}

func (this *MapSum) Sum(prefix string) int {
	sum := 0
	idx := sort.Search(len(this.items), func(i int) bool {
		return this.items[i].key >= prefix
	})
	for i := idx; i < len(this.items); i++ {
		if !strings.HasPrefix(this.items[i].key, prefix) {
			break
		}
		sum += this.items[i].value
	}
	return sum
}
